cmd/git_utils.go

```diff
@@ -99,31 +103,67 @@ func GetGitInfo(config *RootCommandConfig) (GitInfo, error) {
 
 	config.Debug.log("git version: ", version)
 
-	kargs := []string{"status", "-sb"}
-
-	output, gitErr := RunGit(config.LoggingConfig, config.ProjectDir, kargs, config.Dryrun)
+	gitInfo.Commit, gitErr = RunGitGetLastCommit(gitInfo.RemoteURL, config)
 	if gitErr != nil {
-		return gitInfo, gitErr
+		errMsg += "Received error getting current commit: " + gitErr.Error()
 	}
 
 	lineSeparator := "\n"
 	if runtime.GOOS == "windows" {
 		lineSeparator = "\r\n"
 	}
-	output = strings.Trim(output, trimChars)
-	outputLines := strings.Split(output, lineSeparator)
+
+	outputLines, err := RunGitBranchContains(config.LoggingConfig, gitInfo.Commit.SHA, config.ProjectDir, lineSeparator, config.Dryrun)
+	if err != nil {
+		errMsg += "Unable to locate latest commit in remote. "
+	} else {
+		gitInfo.Upstream = outputLines[0]
+		for _, upstream := range outputLines {
+			if strings.Contains(upstream, "origin") {
+				gitInfo.Upstream = upstream
+				break
+			} else if strings.Contains(upstream, "upstream") {
+				gitInfo.Upstream = upstream
+			}
+		}
+		gitInfo.Upstream = strings.TrimSpace(gitInfo.Upstream)
+	}
+
+	if gitInfo.Upstream == "" {
+		gitInfo.Commit.Pushed = false
+		gitRemoteOutput, err := RunGitRemote(config.LoggingConfig, config.ProjectDir, lineSeparator, config.Dryrun)
+		if err != nil {
+			return gitInfo, err
+		}
+
+		for _, remotes := range gitRemoteOutput {
+			if strings.Contains(remotes, "origin") {
+				gitInfo.Upstream = remotes
+				break
+			} else if strings.Contains(remotes, "upstream") {
+				gitInfo.Upstream = remotes
+			}
+		}
+	}
 
 	const noCommits = "## No commits yet on "
 	const branchPrefix = "## "
 	const branchSeparatorString = "..."
 
-	value := strings.Trim(outputLines[0], trimChars)
+	kargs := []string{"status", "-sb"}
+	statusOutput, gitErr := RunGit(config.LoggingConfig, config.ProjectDir, kargs, config.Dryrun)
+	if gitErr != nil {
+		return gitInfo, gitErr
+	}
+
+	statusOutput = strings.Trim(statusOutput, trimChars)
+	statusOutputLines := strings.Split(statusOutput, lineSeparator)
+
+	value := strings.Trim(statusOutputLines[0], trimChars)
```

Comment:
I would still run `git status -sb` first, and then go to `branch contains` and check that the `origin/master` appears because it could be a remote which is not `origin` or `upstream` or the 1st.

Reply:
Okay, so we should still keep the original method of getting the upstream from `git status -sb` and only resorting to the other methods if we can't retrieve it from there?

Reply:
Yup, since for now we are not adding the commit url

Reply:
Cool, added in the latest commit :)